src/Pos/Block/Logic.hs

```diff
@@ -147,6 +149,34 @@ lcaWithMainChain headers =
             ([], True)   -> pure $ Just h
             (x:xs, True) -> lcaProceed (Just h) (x :| xs)
 
+-- | The phrase “we're in recovery mode” is confusing because it can mean two
+-- different things:
+--
+-- 1. Last known block is more than K slots away from the current slot, or
+--    current slot isn't known.
+--
+-- 2. We're actually in the process of requesting blocks because we have
+--    detected that #1 happened. (See 'ncRecoveryHeader' and
+--    'recoveryInProgress'.)
+--
+-- This function checks for #1. Note that even if we're doing recovery right
+-- now, 'needRecovery' will still return 'True'.
+--
+needRecovery
+    :: forall ssc m.
+       (SscWorkersClass ssc, WorkMode ssc m)
+    => m Bool
+needRecovery =
+    getCurrentSlot >>= \case
+        Nothing   -> pure True
+        Just slot -> isTooOld slot
+  where
+    isTooOld slot = do
+        knownSlot <- getEpochOrSlot <$> DB.getTipBlockHeader @ssc
+        pure $
+            (knownSlot & from enum %~ (+ blkSecurityParam)) <
```

Comment:
I will clarify: _for me_ it is weird and hard to see. I.e. unless I look closely it looks like a weird `--`.

If someone feels strongly about it, I'll change it.